### PR TITLE
Secret providers

### PR DESCRIPTION
**Pull request by GunaKKIBM**

### CODE REVIEW

File: pkg/k8s_utils/secret.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+	// Creating k8s client used to read secret
+	clientset, err := kubernetes.NewForConfig(k8sConfig)
+	if err != nil {
+		logger.Error("Error creating k8s client", zap.Error(err))
+		return "", utils.Error{Description: fmt.Sprintf("Error fetching secret %s - unable to create k8s client", secretname), BackendError: err.Error()}
+	}
+
+	// Reading the namespace in which the pod is deployed
+	byteData, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
```
Comment (arahamad) [inline pkg/k8s_utils/secret.go:52]: why this is hardcoded ?

**Review by arahamad — CHANGES_REQUESTED** [review]
can you check this comment

File: secretprovider/secretprovider.proto [hunk 2]
```diff
@@ -17,6 +17,7 @@ service SecretProvider {
 message Request {
   string secret = 1;
   bool 	 isFreshTokenRequired = 2;
+  bool   readSecret = 3;
```
Comment (GunaKKIBM) [inline secretprovider/secretprovider.proto:20]: Remove this readSecret

File: pkg/authenticator/authenticator.go [hunk 3]
```diff
@@ -63,46 +67,30 @@ func NewAuthenticator(logger *zap.Logger) (Authenticator, string, error) {
 		return authenticator, credentialType, nil
 	}
 
-	logger.Error("IBMCLOUD_CREDENTIALS_FILE undefined, trying to read storage secret store")
-	conf, err := config.ReadConfig(logger)
+	// Parse it the secret is storage-secret-store
```
Comment (arahamad) [inline pkg/authenticator/authenticator.go:70]: Can we have default case as well in switch case at line 64 and return the error

File: pkg/authenticator/authenticator.go [hunk 4]
```diff
@@ -117,14 +105,14 @@ func parseCredentials(logger *zap.Logger, credentialFilePath string) (map[string
...
-		logger.Error("Credentials provided are not in the expected format", zap.String("Credentials file path", credentialFilePath))
+		logger.Error("Credentials provided are not in the expected format")
 		return nil, utils.Error{Description: utils.ErrInvalidCredentialsFormat}
 	}
 
 	// validating credentials
 	credentialType, ok := credentialsmap[utils.IBMCLOUD_AUTHTYPE]
 	if !ok {
-		logger.Error("IBMCLOUD_AUTHTYPE is undefined", zap.String("Credentials file path", credentialFilePath))
+		logger.Error("IBMCLOUD_AUTHTYPE is undefined")
```
Comment (arahamad) [inline pkg/authenticator/authenticator.go:115]: Can we have more detail error, applicable for other places as well

File: pkg/config/config.go [hunk 5]
```diff
@@ -136,35 +132,22 @@ type APIConfig struct {
...
+		logger.Error("Failed to parse config", zap.Error(err))
+		return nil, utils.Error{Description: "Failed to parse config", BackendError: err.Error()}
 	}
 
-	err = envconfig.Process("", conf)
+	err = envconfig.Process("", configData)
 	if err != nil {
 		logger.Error("Failed to gather environment config variable", zap.Error(err))
-		return utils.Error{Description: "Failed to gather environment variables", BackendError: err.Error()}
+		return nil, utils.Error{Description: "Failed to gather environment variables", BackendError: err.Error()}
```
Comment (arahamad) [inline pkg/config/config.go:148]: Can we add this error code in the message file, applicable for other places as well

File: secrets/storage-secret-store/slclient.toml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  iam_client_id = "bx"
+  iam_client_secret = "bx"
+  iam_api_key = ""
+  refresh_token = ""
+  pay_tier = "paid"
+  containers_api_route = "https://us-south.containers.cloud.ibm.com"
+  encryption = true
+  containers_api_route_private = "https://private.us-south.containers.cloud.ibm.com"
+
+[Softlayer]
```
Comment (arahamad) [inline secrets/storage-secret-store/slclient.toml:12]: can we remove this section

**Review by arahamad — CHANGES_REQUESTED** [review]
Can you please check these comments

File: pkg/authenticator/authenticator.go (hunk 3 above)
Comment (GunaKKIBM) [reply pkg/authenticator/authenticator.go:70]: This is taken care of in the parseIBMCloudCredentials function, where AUTH_TYPE is validated. If there is any other AUTH_TYPE, that function will return error

File: pkg/k8s_utils/secret.go (hunk 1 above)
Comment (GunaKKIBM) [reply pkg/k8s_utils/secret.go:52]: As discussed, this will be used. the path will be moved to constants file

**Review by arahamad — APPROVED** [review]

**Pull request merged**